dodo.py

```diff
@@ -0,0 +1,400 @@
+    """Push all tags for a Jupyter image - only should be done after they have been tested"""
+    for image in P.ALL_IMAGES:
+        yield dict(
+            name=f"push:{image}",
+            doc="Push the image to the specified registry",
+            actions=[
+                U.do(
+                    "echo",
+                    f"::group::Push {U.registry_image(registry, image)} system's architecture",
+                ),
+                U.do(
+                    "docker",
+                    "push",
+                    "--all-tags",
+                    f"{U.registry_image(registry, image)}",
+                ),
+                U.do("echo", "::endgroup::"),
+            ],
+        )
+
+
+# -----------------------------------------------------------------------------
+# Support classes and methods
+# -----------------------------------------------------------------------------
+
+
+class P:
+    """Paths to project files and directories, used to provide consistency across the multiple doit tasks"""
+
+    DODO = Path(__file__)
+    ROOT = DODO.parent
+
+    # docs
+    DOCS = ROOT / "docs"
+    DOCS_TARGET = ROOT / "docs/_build/html"
+    README = ROOT / "README.md"
+    DOCS_PY = sorted(DOCS.rglob("*.py"))
+    DOCS_RST = sorted(DOCS.rglob("*.rst"))
+    DOCS_SRC_MD = sorted(DOCS.rglob("*.md"))
+    DOCS_MD = sorted([*DOCS_SRC_MD, README])
+
+    # wiki
+    WIKI = ROOT.parent / "wiki"
+    WIKI_MANIFEST = WIKI / "manifests"
+
+    # tests
+    TESTS = ROOT / "tests"
+    TESTS_RUN = TESTS / "run_tests.py"
+
+    # CI
+    CI = ROOT / ".github"
+    CI_IMG = CI / "built-docker-images"
+
+    # Docker-related
+    OWNER = "jupyter"
+    DOCKER_REGISTRY = "dockerhub"
+
+    # Images supporting the following architectures:
+    # - linux/amd64
+    # - linux/arm64
+    MULTI_IMAGES = [
+        "base-notebook",
+        "minimal-notebook",
+        "scipy-notebook",
+        "r-notebook",
+        "pyspark-notebook",
+        "all-spark-notebook",
+    ]
+
+    # TODO: uncomment after initial tests
+    # ALL_IMAGES = [
+    #     "base-notebook",
+    #     "minimal-notebook",
+    #     "scipy-notebook",
+    #     "r-notebook",
+    #     "tensorflow-notebook",
+    #     "datascience-notebook",
+    #     "pyspark-notebook",
+    #     "all-spark-notebook",
+    # ]
+    ALL_IMAGES = [
+        "base-notebook",
+        "minimal-notebook",
+    ]
+
+    AMD64_IMAGES = ["datascience-notebook", "tensorflow-notebook"]
+
+
+class U:
+    """Supporting methods and variables"""
+
+    # CI specific
+    IS_CI = bool(os.environ.get("CI", 0))
+
+    # args
+    PYM = ["python", "-m"]
+    PYTEST_ARGS = ["pytest", "-m", "not info"]
+
+    # git specific - used for tagging
+    SOURCE_DATE_EPOCH = (
+        subprocess.check_output(["git", "log", "-1", "--format=%ct"])
+        .decode("utf-8")
+        .strip()
+    )
+
+    GIT_COMMIT_SHA = (
+        subprocess.check_output(["git", "rev-parse", "--short", "HEAD"])
```

Comment:
Maybe we could use plumbum in places like this to make our life easier?
An example: https://github.com/jupyter/docker-stacks/blob/master/tagging/git_helper.py

Reply:
I will have to look at how these two work together - when I tried to call the tests directly as in the Makefile I kept getting errors since plumbum is used under the hood

Reply:
I directly used tagging.git_helper and it works well.